manpages/makeivs-ng.1
.TH MAKEIVS-NG 1 "December 2018" "Version 1.5.2"

.SH NAME
makeivs - generate a dummy IVS dump file with a specific WEP key
.SH SYNOPSIS
.B makeivs
<ivs file> <104-bit key>
.SH DESCRIPTION
.BI makeivs-ng
is a tool designed to generate an IVS dump file with an inputted WEP key.
The aim of is tools is to provide a way to create dumps with a known encryption key for tests.
.SH OPTIONS
.TP
.B Common options:
.TP
.I -b <bssid> or --bssid <bssid>
Set the BSSID (Access Point MAC)
.TP
.I -f <num> or --first <num>
Value for the first IV generated.
.TP
.I -k <key> or --key <key>
Target network WEP key in hex. Separator between bytes is accepted but not necessary.
.TP
.I -s <num> or --seed <num>
Seed used to setup random generator. May be used in combination with \-p or \-\-prng.
.TP
.I -w <file> or --write <file>
Filename to write IVs into.
.TP
.I -c <num> or --count <num>
Amount of IVs to generate. Default value is 100000.
.TP
.I -d <num> or --dupe <num>
Percentage of duplicate IVs.
.TP
.I -e <num> or --error <num>
Percentage of erroneous keystreams.
.TP
.I -l <num> or --length <num>
Size of keystreams. Default: 16 bytes.
.TP
.I -n or --nofms 
Ignores weak IVs.
.TP
.I -p or --prng
Use random values when generating IVs. Default is to use sequential values.
.TP
.I --help
Show help screen.
.SH EXAMPLE
.B makeivs
makeivs -w out.ivs -k 123456789ABCDEF123456789AB
.SH AUTHOR
This manual page was written by Adam Cecile <gandalf@le-vert.net> for the Debian system (but may be used by others).
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or any later version published by the Free Software Foundation
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.
.SH SEE ALSO
.br
.B airbase-ng(8)
.br
.B aireplay-ng(8)
.br
.B airmon-ng(8)
.br
.B airodump-ng(8)
.br
.B airodump-ng-oui-update(8)
.br
.B airserv-ng(8)
.br
.B airtun-ng(8)
.br
.B besside-ng(8)
.br
.B easside-ng(8)
.br
.B tkiptun-ng(8)
.br
.B wesside-ng(8)
.br
.B aircrack-ng(1)
.br
.B airdecap-ng(1)
.br
.B airdecloak-ng(1)
.br
.B airolib-ng(1)
.br
.B besside-ng-crawler(1)
.br
.B buddy-ng(1)
.br
.B ivstools(1)
.br
.B kstats(1)
.br
.B packetforge-ng(1)
.br
.B wpaclean(1)
.br
.B airventriloquist(8)


CHANGELOG
Changelog for 2.0.9
-fixed client and clientless attacks chopchop and frag packet injection not generating IVs because -h MAC was not set
-force frag, chopchop and aircrack-ng xterms to stay on screen after successfull command so you can see results
-cleaned the code some more
-cleaning functions needs more tweaks
-atheros users need to use options 9 (monitor mode) before other commands to create-destroys correct VAPS
-fixed "unary operator expected" msg (harmless bug) (not fully fixed because of "" generating errors in logic tests)
-tweaked 2 steps attacks (frag, chopchop) to go to menu to inject after successfull first step
-added option to jump to injection menu (added this to the main menu)
-fixed AP encryption detection for cracking step (bug was introduced by "")
-lowered adapter speed to 1M for wesside-ng attacks
-added -hold to mdk attack to allow users to see results before the windows closes

Changelog for 2.0.8 (27/01/2008)
-added wesside-ng airoscript style usage
-added interface type detection
-tweaked menu
-first attempt at making airoscript work with different adapters (add more card support)(BIG THX TO GREEN-FREQ!!!)
-atheros device are now supported
-added mdk3
-added back to main menu options
-fixed chopchop attack (BIG THX TO GREEN-FREQ!!!)
-tweaked FAKE_MAC function for atheros

Changelog for 2.0.7 BETA1 (21/01/2007)
-changed capture from IVs to full date (ivs to cap)
-updated script for new implementation of fragmentation attack
-other minor modifications and tweaks

Changelog for PRE-1.8.2-BETA1 (30/11/2006)
-fixed all commands leading to : "head: Binary: invalid number of lines" (MAJOR THX TO ANDREA)
-fixed all commands where SSID was truncated (MAJOR THX TO PILOTSNIPES)

Changelog for 1.8.1 (03/11/2006)
-bug fix release
-fixed some problem in the SSID detection function (MAJOR THX TO andreahaku)
-cleaned a few more typos and bugs
-documentation completed a bit

Changelog for 1.8 (17/10/2006)
-dropped support for aircrack-ng 0.6.1 and older
-will only work with patched version of aircrack-ng 0.6.2 (patches can be found in the forum under the programming section)
-specifique channels hopping (for example channel 1,6,11 or 1-3,5-7,9,11)
-Access point filtering during attack (airodump will only list attacked Access point)
-new SSID detection, works with ALL ssid (supports all character and spacing possible)
-new Access point selection menu that list all AP with encryption type (eliminates the WEP? problem)
-option 2) now get you directly to a full list of AP
-chopchop attack and fragmentation attack implement
-WPA dictionnary attack implemented
-changed menu 3,4,5 based on detection of AP encryption
-added option to jump to associated client list from ap selection
-color change support for xterm
-Windows size and placement fully changeable from variables settings
-changed filenames from SSID to MAC because of possible issues if weird SSID (to eliminate security menace (if SSID is something like "rm -rf /*" etc...)
-added option an option to run "1) scan 2) select 3) attack" one after the other
-added some comments in the script to help users
-added greetings message (can be disabled setting WELCOME="1" to "0")

Changelog for 1.7RC7 (24/09/2006)
-added some resilience to the configure function now works properly with hidden ssid's and rausb devices 

Changelog for 1.7RC6 (19/09/2006)
-Code cleanup
-added deauth command to solointeractiveattack function
-changed base variables (fudge factor & inject rate)

Changelog for 1.7RC5 (14/09/2006)
-Fixed minor cosmetic issue caused by carriage return char when ssid is missing (15/9/06)
-Fixed the weird character match issue!! edit at will!!!
-added an abitrary variable ($acv) to the aircrackversioncheck function which can be used in future to keep the script backwards compatible (had to guess at what version 0.6.0 of aircrack-ng displays when run on it's own? and version 0.5.so might not work??)
-used new variable in crack function to check version and run correct crack command for the aricrack version installed fixing the bug on the B|T forum.

Changelog for 1.7RC4 (13/09/2006)
-corrected minor speellling issues ;o) 
-added aircrack-ng version detection function currently all it does is display aircrack version in the menu function (should be useful when new functionality is added to aircrack to allow backwards compatibility)
-added solointeractiveattack function (designed to run just the airreplay interactive attack function only for use with an attack that has stalled i.e. data has stopped increasing, the user can try this for another injection thread! (might not actually work?? needs more testing does spawn and seems to work!)
-added option to manually add ssid if a blank ssid is detected (this has some problems when certain editors are used as this is not strictly blank but a non printable character recommend kwrite on Back|track to edit script)


src/makeivs-ng.c
/*
  * Server for osdep network driver.  Uses osdep itself!  [ph33r teh recursion]
  *
  *  Copyright (C) 2006-2018 Thomas d'Otreppe <tdotreppe@aircrack-ng.org>
  *  Copyright (C) 2004, 2005 Christophe Devine
  *
  *  This program is free software; you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
  *  the Free Software Foundation; either version 2 of the License, or
  *  (at your option) any later version.
  *
  *  This program is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  *  GNU General Public License for more details.
  *
  *  You should have received a copy of the GNU General Public License
  *  along with this program; if not, write to the Free Software
  *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <time.h>

#include "version.h"
#include "pcap.h"
#include "uniqueiv.h"
#include "aircrack-util/common.h"

#define NULL_MAC "\x00\x00\x00\x00\x00\x00"

char usage[] =

	"\n"
	"  %s - (C) 2006-2018 Thomas d\'Otreppe\n"
	"  https://www.aircrack-ng.org\n"
	"\n"
	"  usage: makeivs-ng [options]\n"
	"\n"
	"  Common options:\n"
	"      -b <bssid> : Set access point MAC address\n"
	"      -f <num>   : Number of first IV\n"
	"      -k <key>   : Target network WEP key in hex\n"
	"      -s <num>   : Seed used to setup random generator\n"
	"      -w <file>  : Filename to write IVs into\n"
	"      -c <num>   : Number of IVs to generate\n"
	"      -d <num>   : Percentage of dupe IVs\n"
	"      -e <num>   : Percentage of erroneous keystreams\n"
	"      -l <num>   : Length of keystreams\n"
	"      -n         : Ignores weak IVs\n"
	"      -p         : Uses prng algorithm to generate IVs\n"
	"\n"
	"      --help     : Displays this usage screen\n"
	"\n";

int main(int argc, char * argv[])
{
	int i, j, k, pre_n, n, count = 100000, length = 16;
	int paramUsed = 0, keylen = 0, zero = 0, startiv = 0, iv = 0;
	FILE * f_ivs_out;
	unsigned char K[32];
	unsigned char S[256];
	//     unsigned char buffer[64];
	char *s, *filename = NULL;
	struct ivs2_pkthdr ivs2;
	struct ivs2_filehdr fivs2;
	unsigned long long size;
	int option_index, option, crypt = 0;
	char buf[2048];
	int weplen = 0, nofms = 0, prng = 0;
	float errorrate = 0, dupe = 0;
	unsigned char bssid[6];
	int seed = time(NULL), z;
	int maxivs = 0x1000000;
	unsigned char byte;
	unsigned char ** uiv_root;

	static struct option long_options[] = {{"key", 1, 0, 'k'},
										   {"write", 1, 0, 'w'},
										   {"count", 1, 0, 'c'},
										   {"seed", 1, 0, 's'},
										   {"length", 1, 0, 'l'},
										   {"first", 1, 0, 'f'},
										   {"bssid", 1, 0, 'b'},
										   {"dupe", 1, 0, 'd'},
										   {"error", 1, 0, 'e'},
										   {"nofms", 0, 0, 'n'},
										   {"prng", 0, 0, 'p'},
										   {"help", 0, 0, 'H'},
										   {0, 0, 0, 0}};

	i = 0;
	memset(K, 0, 32);
	memset(bssid, 0, 6);
	uiv_root = uniqueiv_init();

	/* check the arguments */

	do
	{
		option_index = 0;

		option = getopt_long(
			argc, argv, "k:w:c:s:l:f:b:d:e:npHh", long_options, &option_index);

		if (option < 0) break;

		switch (option)
		{
			case 0:

				break;

			case ':':

				goto usage;
				break;

			case '?':

				goto usage;
				break;

			case 'n':

				paramUsed = 1;
				nofms = 1;
				break;

			case 'p':

				paramUsed = 1;
				prng = 1;
				break;

			case 'l':

				paramUsed = 1;
				if (atoi(optarg) < 2 || atoi(optarg) > 2300)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));
					printf("Specified keystream length is invalid. [2-2300]");
					return (1);
				}

				length = atoi(optarg);
				break;

			case 'c':

				paramUsed = 1;
				if (atoi(optarg) < 1 || atoi(optarg) > 0x1000000)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));
					printf("Specified number of IVs is invalid. [1-16777216]");
					return (1);
				}

				count = atoi(optarg);
				break;

			case 's':

				paramUsed = 1;
				if (atoi(optarg) < 1)
				{
					printf("Specified seed is invalid. [>=1]");
					printf("\"%s --help\" for help.\n", argv[0]);
					return (1);
				}

				seed = atoi(optarg);
				break;

			case 'e':

				paramUsed = 1;
				sscanf(optarg, "%f", &errorrate);
#if defined(__x86_64__) && defined(__CYGWIN__)
				if (errorrate < 0.0f || errorrate > (0.0f + 100))
				{
#else
				if (errorrate < 0.0f || errorrate > 100.0f)
				{
#endif
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));
					printf("Specified errorrate is invalid. [0-100]");
					return (1);
				}

				break;

			case 'd':

				paramUsed = 1;
				if (sscanf(optarg, "%f", &dupe) != 1 || dupe < 0.0f
					|| dupe >
#if defined(__x86_64__) && defined(__CYGWIN__)
						   (0.0f + 100))
				{
#else
						   100.0f)
				{
#endif
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));

					printf("Specified dupe is invalid. [0-100]");
					return (1);
				}

				break;

			case 'f':

				paramUsed = 1;
				if (atoi(optarg) < 0 || atoi(optarg) > 0xFFFFFF)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));

					printf("Specified start IV is invalid. [0-16777215]");
					return (1);
				}

				paramUsed = 1;
				startiv = atoi(optarg);
				break;

			case 'w':

				paramUsed = 1;
				filename = optarg;
				break;

			case 'b':

				paramUsed = 1;
				if (memcmp(bssid, NULL_MAC, 6) != 0)
				{
					printf("Notice: bssid already given\n");
					break;
				}
				if (getmac(optarg, 1, bssid) != 0)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));

					printf("Notice: invalid bssid\n");
					return (1);
				}
				break;

			case 'k':

				paramUsed = 1;
				if (crypt != 0)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));

					printf("Encryption key already specified.\n");
					return (1);
				}

				crypt = 1;

				i = 0;
				s = optarg;

				buf[0] = s[0];
				buf[1] = s[1];
				buf[2] = '\0';

				while (sscanf(buf, "%x", &n) == 1)
				{
					if (n < 0 || n > 255)
					{
						printf(usage,
							   getVersion("makeivs-ng",
										  _MAJ,
										  _MIN,
										  _SUB_MIN,
										  _REVISION,
										  _BETA,
										  _RC));

						printf("Invalid WEP key.\n");
						return (1);
					}

					if (3 + i >= 32) break;

					K[3 + i++] = n;

					s += 2;

					if (s[0] == ':' || s[0] == '-') s++;

					if (s[0] == '\0' || s[1] == '\0') break;

					buf[0] = s[0];
					buf[1] = s[1];
				}

				if (i != 5 && i != 13 && i != 29)
				{
					printf(usage,
						   getVersion("makeivs-ng",
									  _MAJ,
									  _MIN,
									  _SUB_MIN,
									  _REVISION,
									  _BETA,
									  _RC));

					printf("Invalid WEP key length. [5,13,29]\n");
					return (1);
				}

				weplen = i;
				keylen = i + 3;

				break;

			case 'h':
			case 'H':
				goto usage;
				break;

			default:
				goto usage;
		}
	} while (1);

	if (nofms) maxivs -= 256 * weplen;

	srand(seed);

	if (paramUsed == 0)
	{
	usage:
		printf(usage,
			   getVersion(
				   "makeivs-ng", _MAJ, _MIN, _SUB_MIN, _REVISION, _BETA, _RC));
		return (0);
	}

	if (count > maxivs)
	{
		printf(usage,
			   getVersion(
				   "makeivs-ng", _MAJ, _MIN, _SUB_MIN, _REVISION, _BETA, _RC));

		printf("Specified too many IVs (%d), but there are only %d possible.\n",
			   count,
			   maxivs);
		return (1);
	}

	if (length == 0) length = 16; // default 16 keystreambytes

	if (crypt < 1)
	{
		printf(usage,
			   getVersion(
				   "makeivs-ng", _MAJ, _MIN, _SUB_MIN, _REVISION, _BETA, _RC));

		printf("You need to specify the WEP key (-k).\n");
		return (1);
	}

	if (filename == NULL)
	{
		printf(usage,
			   getVersion(
				   "makeivs-ng", _MAJ, _MIN, _SUB_MIN, _REVISION, _BETA, _RC));

		printf("You need to specify the output filename (-w).\n");
		return (1);
	}

	size = (long long) strlen(IVS2_MAGIC)
		   + (long long) sizeof(struct ivs2_filehdr)
		   + (long long) count * (long long) sizeof(struct ivs2_pkthdr)
		   + (long long) count * (long long) (length + 4);

	printf("Creating %d IVs with %d bytes of keystream each.\n", count, length);
	printf("Estimated filesize: ");
	if (size > 1024 * 1024 * 1024) // over 1 GB
		printf("%.2f GB\n", ((double) size / (1024.0 * 1024.0 * 1024.0)));
	else if (size > 1024 * 1024) // over 1 MB
		printf("%.2f MB\n", ((double) size / (1024.0 * 1024.0)));
	else if (size > 1024) // over 1 KB
		printf("%.2f KB\n", ((double) size / 1024.0));
	else // under 1 KB
		printf("%.2f Byte\n", (double) size);

	if ((f_ivs_out = fopen(filename, "wb+")) == NULL)
	{
		perror("fopen");
		return (1);
	}

	if (fwrite(IVS2_MAGIC, 1, 4, f_ivs_out) != (size_t) 4)
	{
		perror("fwrite(IVs file MAGIC) failed");
		return (1);
	}

	memset(&fivs2, '\x00', sizeof(struct ivs2_filehdr));
	fivs2.version = IVS2_VERSION;

	/* write file header */
	if (fwrite(&fivs2, sizeof(struct ivs2_filehdr), 1, f_ivs_out) != (size_t) 1)
	{
		perror("fwrite(IV file header) failed");
		return (1);
	}

	memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
	ivs2.flags |= IVS2_BSSID;
	ivs2.len += 6;

	/* write header */
	if (fwrite(&ivs2, 1, sizeof(struct ivs2_pkthdr), f_ivs_out)
		!= (size_t) sizeof(struct ivs2_pkthdr))
	{
		perror("fwrite(IV header) failed");
		return (1);
	}

	if (memcmp(NULL_MAC, bssid, 6) == 0)
	{
		memcpy(bssid, "\x01\x02\x03\x04\x05\x06", 6);
	}

	/* write BSSID */
	if (fwrite(bssid, 1, 6, f_ivs_out) != (size_t) 6)
	{
		perror("fwrite(IV bssid) failed");
		return (1);
	}
	printf("Using fake BSSID %02X:%02X:%02X:%02X:%02X:%02X\n",
		   bssid[0],
		   bssid[1],
		   bssid[2],
		   bssid[3],
		   bssid[4],
		   bssid[5]);
	z = 0;
	pre_n = 0;
	for (n = 0; n < count; n++)
	{
		if ((dupe == 0) || (pre_n == n)
			|| ((float) rand() / (float) RAND_MAX > (float) ((float) dupe /
#if defined(__x86_64__) && defined(__CYGWIN__)
															 (0.0f + 100))))
#else
															 100.0f)))
#endif
		{
			if (prng)
			{
				iv = rand() & 0xFFFFFF;
			}
			else
			{
				iv = (z + startiv) & 0xFFFFFF;
				z++;
			}

			if (nofms)
			{
				if ((iv & 0xff00) == 0xff00)
				{
					byte = (iv >> 16) & 0xff;
					if (byte >= 3 && byte < keylen)
					{
						if (!prng && (iv & 0xFF) == 0) z += 0xff;
						n--;
						continue;
					}
				}
			}

			if (uniqueiv_check(uiv_root, (unsigned char *) &iv) != 0)
			{
				n--;
				continue;
			}

			uniqueiv_mark(uiv_root, (unsigned char *) &iv);
		}

		pre_n = n;

		K[2] = (iv >> 16) & 0xFF;
		K[1] = (iv >> 8) & 0xFF;
		K[0] = (iv) &0xFF;

		for (i = 0; i < 256; i++) S[i] = i;

		for (i = j = 0; i < 256; i++)
		{
			j = (j + S[i] + K[i % keylen]) & 0xFF;
			SWAP(S[i], S[j]);
		}

		if (errorrate > 0 && ((float) ((float) rand() / (float) RAND_MAX)
							  <= (float) (errorrate /
#if defined(__x86_64__) && defined(__CYGWIN__)
										  (0.0f + 100))))
#else
										  100.0f)))
#endif
		{
			SWAP(S[1], S[11]);
		}

		memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
		ivs2.flags = 0;
		ivs2.len = 0;

		ivs2.flags |= IVS2_XOR;
		ivs2.len += length + 4;

		if (fwrite(&ivs2, 1, sizeof(struct ivs2_pkthdr), f_ivs_out)
			!= (size_t) sizeof(struct ivs2_pkthdr))
		{
			perror("fwrite(IV header) failed");
			return (1);
		}

		if (fwrite(K, 1, 3, f_ivs_out) != (size_t) 3)
		{
			perror("fwrite(IV iv) failed");
			return (1);
		}
		if (fwrite(&zero, 1, 1, f_ivs_out) != (size_t) 1)
		{
			perror("fwrite(IV idx) failed");
			return (1);
		}
		ivs2.len -= 4;

		i = j = 0;
		for (k = 0; k < length; k++)
		{
			i = (i + 1) & 0xFF;
			j = (j + S[i]) & 0xFF;
			SWAP(S[i], S[j]);
			fprintf(f_ivs_out, "%c", S[(S[i] + S[j]) & 0xFF]);
		}
		if ((n % 10000) == 0)
			printf("%2.1f%%\r",
				   ((float) n / (float) count) *
#if defined(__x86_64__) && defined(__CYGWIN__)
					   (0.0f + 100));
#else
					   100.0f);
#endif
		fflush(stdout);
	}

	fclose(f_ivs_out);
	printf("Done.\n");
	return (0);
}
